### PR TITLE
[receiver/carbon] Permit float timestamps

### PR DESCRIPTION
**Pull request by jonnangle**
**Description:** 

This allows timestamps to be sent as floats.

**Link to tracking Issue:** #31312 

**Testing:** 

Added unit test case with float timestamp.

### CODE REVIEW

File: receiver/carbonreceiver/protocol/plaintext_parser_test.go [hunk 1]
```diff
@@ -41,6 +41,16 @@ func Test_plaintextParser_Parse(t *testing.T) {
...
+		{
+			line: "tst.floattimestamp 3.14 1582230020.123",
+			want: buildDoubleMetric(
+				GaugeMetricType,
+				"tst.floattimestamp",
+				nil,
+				1582230020,
+				3.14,
+			),
+		},
```
Comment (aboguszewski-sumo) [inline receiver/carbonreceiver/protocol/plaintext_parser_test.go:53]: Did you confirm that this particular test fails before the fix?

**Review by aboguszewski-sumo — APPROVED** [review]
One question in the review, besides that LGTM.
Also, please authorise yourself.

File: receiver/carbonreceiver/protocol/plaintext_parser_test.go (hunk 1 above)
Comment (jonnangle) [reply receiver/carbonreceiver/protocol/plaintext_parser_test.go:53]: yes, without the fix:

```
=== FAIL: protocol Test_plaintextParser_Parse/tst.floattimestamp_3.14_1582230020.123 (0.00s)
    plaintext_parser_test.go:101:
        	Error Trace:	/Users/jon.nangle/src/opentelemetry-collector-contrib/receiver/carbonreceiver/protocol/plaintext_parser_test.go:101
        	Error:      	Not equal:
        	            	expected: pmetric.Metric{orig:(*v1.Metric)(0xc0001a5fc0), state:(*internal.State)(0xc000264cdc)}
        	            	actual  : pmetric.Metric{orig:(*v1.Metric)(nil), state:(*internal.State)(nil)}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,24 +1,4 @@
        	            	 (pmetric.Metric) {
        	            	- orig: (*v1.Metric)({
        	            	-  Name: (string) (len=18) "tst.floattimestamp",
        	            	-  Description: (string) "",
        	            	-  Unit: (string) "",
        	            	-  Data: (*v1.Metric_Gauge)({
        	            	-   Gauge: (*v1.Gauge)({
        	            	-    DataPoints: ([]*v1.NumberDataPoint) (len=1) {
        	            	-     (*v1.NumberDataPoint)({
        	            	-      Attributes: ([]v1.KeyValue) <nil>,
        	            	-      StartTimeUnixNano: (uint64) 0,
        	            	-      TimeUnixNano: (uint64) 1582230020000000000,
        	            	-      Value: (*v1.NumberDataPoint_AsDouble)({
        	            	-       AsDouble: (float64) 3.14
        	            	-      }),
        	            	-      Exemplars: ([]v1.Exemplar) <nil>,
        	            	-      Flags: (uint32) 0
        	            	-     })
        	            	-    }
        	            	-   })
        	            	-  })
        	            	- }),
        	            	- state: (*internal.State)(0)
        	            	+ orig: (*v1.Metric)(<nil>),
        	            	+ state: (*internal.State)(<nil>)
        	            	 }
        	Test:       	Test_plaintextParser_Parse/tst.floattimestamp_3.14_1582230020.123
    plaintext_parser_test.go:102:
        	Error Trace:	/Users/jon.nangle/src/opentelemetry-collector-contrib/receiver/carbonreceiver/protocol/plaintext_parser_test.go:102
        	Error:      	Not equal:
        	            	expected: false
        	            	actual  : true
        	Test:       	Test_plaintextParser_Parse/tst.floattimestamp_3.14_1582230020.123

=== FAIL: protocol Test_plaintextParser_Parse (0.00s)
```

**Review by fatsheep9146 — APPROVED** [review]

**Comment by jonnangle** [comment]
> One question in the review, besides that LGTM. Also, please authorise yourself.

Authorisation is now in place - thanks @aboguszewski-sumo for reviewing so quickly!

**Comment by jonnangle** [comment]
@aboguszewski-sumo we've run this in production today, and it works but has moved some of the errors downstream - we have some clients sending sub-second metrics with different values, which prometheus is then rejecting. 

I'm thinking that truncating the timestamp isn't really the right approach here, and what we should be doing is passing the fractional part of the timestamp into the `pmetric.Metric`. If you agree, I can update the PR and tests to reflect this?

**Comment by aboguszewski-sumo** [comment]
@jonnangle My opinion is that we should mimic the original behaviour. First of all, could you confirm that this would work with prometheus in that case? Second of all, what's exactly the rejection reason?

**Comment by jonnangle** [comment]
> First of all, could you confirm that this would work with prometheus in that case? Second of all, what's exactly the rejection reason?

With the PR as it stands, if I push two carbon metrics like this:

```
my_metric 444 1708527107.111
my_metric 555 1708527107.222
```

then they will end up in the pipeline as two metric datapoints, both with timestamp `1708527107`. If these are sent via a prometheusremotewrite exporter, I get back `err="duplicate sample for timestamp"`.

If instead we push the metrics into the pipeline with their original timestamp, then prometheus will accept them and they can be queried - so pushing these metrics:

```
my_metric 444 1708527309.111
my_metric 555 1708527309.222
```

whilst retaining the fractional part of the timestamp leads to this end result in prometheus:

```
{
  "status": "success",
  "data": {
    "resultType": "matrix",
    "result": [
      {
        "metric": {
          "__name__": "my_metric"
        },
        "values": [
          [
            1708527107,
            "444"
          ],
          [
            1708527309.111,
            "444"
          ],
          [
            1708527309.221,
            "555"
          ],
        ]
      }
    ]
  }
}
```

(prometheus has 1ms resolution)

Of course the extent to which people can make use of these data points will be dependent on the tooling that's used to process the data - but they will at least be there.

**Comment by aboguszewski-sumo** [comment]
@jonnangle I get it and I agree.

**Comment by jonnangle** [comment]
👍 it seems to work well for our use case - we've deployed it here and the "duplicate sample for timestamp" errors have dropped away:

image

I'll update the PR today - thanks again!

**Comment by jonnangle** [comment]
@aboguszewski-sumo This has been stable for us in production for a few days. If you're happy, could we look to get this merged? Thanks again for your help!

**Review by aboguszewski-sumo — APPROVED** [review]
LGTM.
@TylerHelmuth this seems ready to merge imo.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@jonnangle please investigate failing CI

**Pull request merged**